opensrp-core/src/main/java/org/smartregister/repository/EventClientRepository.java

```diff
@@ -2346,4 +2356,94 @@ public List<Event> getEventsByTaskIds(Set<String> taskIds) {
                 + event_column.taskId.name()
                 + " IN (" + StringUtils.repeat("?", ",", taskIds.size()) + ")", taskIds.toArray(new String[0]));
     }
+
+    public DuplicateZeirIdStatus cleanDuplicateMotherIds() throws Exception {
+        String username = Context.getInstance().userService().getAllSharedPreferences().fetchRegisteredANM();
+
+        UniqueIdRepository uniqueIdRepository = Context.getInstance().getUniqueIdRepository();
+
+        Map<String, String> duplicates = Context.getInstance().zeirIdCleanupRepository().getClientsWithDuplicateZeirIds();
+        long unusedIdsCount = uniqueIdRepository.countUnUsedIds();
+
+        Timber.d("%d duplicates for provider: %s - %s", duplicates.size(), username, duplicates.toString());
+
+        if (duplicates.size() > 0) {
+            Timber.d(
+                    "%s: %d duplicates for provider: %s - %s\nUnused Unique IDs: %d",
+                    this.getClass().getSimpleName(),
+                    duplicates.size(),
+                    username,
+                    duplicates,
+                    unusedIdsCount
+            );
+        }
+
+        for (Map.Entry<String, String> duplicate : duplicates.entrySet()) {
+            String baseEntityId = duplicate.getKey();
+            String zeirId = duplicate.getValue();
+
+            JSONObject clientJson = getClientByBaseEntityId(baseEntityId);
+            JSONObject identifiers = clientJson.getJSONObject(AllConstants.IDENTIFIERS);
+
+            long unusedIds = uniqueIdRepository.countUnUsedIds();
+            if (unusedIds <= 2) {
+                Timber.e("%s: No more unique IDs available to assign to %s - %s; provider: %s", this.getClass().getSimpleName(), baseEntityId, zeirId, username);
+                android.content.Context applicationContext = CoreLibrary.getInstance().context().applicationContext();
+                applicationContext.startService(new Intent(applicationContext, PullUniqueIdsIntentService.class));
```

Comment:
We can return at this point and retry once the IDs are pulled to avoid depleting all the IDs for the user

Reply:
We found an edge case scenario where in case the duplicates were more than available on device and at the same time duplicates more than 100 (`OPENMRS_UNIQUE_ID_BATCH_SIZE` - batch size for subsequent unique ID fetches), no more IDs would be fetched.

That's why we switched to a synchronous fetch.